articles/designing_saas.rst

```diff
@@ -0,0 +1,244 @@
+
+This used to be an area where having a tenant per database offered the
+most flexibility, at the cost of extra maintenance work from the
+developer(s). You could create separate tables or columns per tenant in
+each database, and manage those differences across time.
+
+If then you wanted to scale your infrastructure to thousands of tenants,
+you’d create a huge table with many string columns (Value0, Value1, …
+Value500). Probably, the best known example of this model is
+`Salesforce’s multi-tenant
+architecture <http://www.developerforce.com/media/ForcedotcomBookLibrary/Force.com_Multitenancy_WP_101508.pdf>`__.
+
+.. figure:: ../images/articles-saas-fig-3.png
+   :alt: Representing data that varies across tenants, traditional method
+
+   Representing data that varies across tenants, traditional method
+
+In this database model, your tables have a preset collection of custom
+columns, labeled in this image as V1, V2, and V3. Dates and Numbers are
+stored as strings in a format such that they can be converted to their
+native types. When you’re storing data associated with a particular
+tenant, you can then use these custom columns and tailor them to each
+tenant’s special needs.
+
+Fortunately, designing your database to account for “flexible” columns
+became significantly easier with the introduction of semi-structured
+data types. PostgreSQL has a rich set of semi-structured data types that
+include `hstore, json, and
+jsonb </blog/2016/07/14/choosing-nosql-hstore-json-jsonb/>`__. You can
+now represent the previous database schema by simply declaring a jsonb
+column and scale to thousands of tenants.
+
+.. figure:: ../images/articles-saas-fig-4.png
+   :alt: Representing data that varies across tenants, modern method
+
+   Representing data that varies across tenants, modern method
+
+Of course, these aren’t the only design criteria and questions to be
+aware of. If you shard your database tables, how do you handle isolation
+or integrate with ORM libraries? What happens if you have a table that
+you can’t easily add a tenant\_id column? In this blog post, we focused
+on building multi-tenant databases with scaling as the primary
+consideration in mind; and skipped over certain points. If you’re
+looking to learn more about designing multi-tenant databases, please
+sign up for our `upcoming webinar on the
+topic! <https://attendee.gotowebinar.com/register/46885752336450820>`__
+
+The good news is, databases have advanced quite a bit in the past ten
+years in accommodating SaaS applications at scale. What was once only
+available to the likes of Google and Salesforce with significant
+engineering effort, is now becoming accessible to everyone with
+open-source technologies such as PostgreSQL and
+`Citus <https://github.com/citusdata/citus>`__. If you’re thinking of
+modernizing your multi-tenant architecture, `drop us a
+line </about/contact_us>`__ and we’d be happy to chat.
+
+.. raw:: html
```

Comment:
I don't think we need the section about signing up for our newsletter in our documentation.